### PR TITLE
[TECH-450] - DB access in EnrollmentInExistingValidationHook

### PR DESCRIPTION
**Pull request by lucaCambi77**
[TECH-450](https://jira.dhis2.org/browse/TECH-450)

Description:

To reduce db access in tracker importer, a map Tracked Entity -> Program Instances has been added in Preheat.
Validation logic related to errors still remains the same. However now we also merge enrollments from payload, in case giving priority to the ones in payload. 

Work done:

- [x] add preheat supplier and create a map  Tracked Entity -> Program Instances
- [x] updates validation hook
- [x] add unit tests for hook
- [x] fix tests

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 1]
```diff
@@ -175,13 +175,13 @@
     private Map<TrackerIdScheme, Map<String, TrackedEntityComment>> notes = new EnumMap<>( TrackerIdScheme.class );
 
     /**
-     * A Map of event uid and preheated {@see ProgramInstance}. The value is a
-     * List, because the system may return multiple ProgramInstance, which will
-     * be detected by validation
+     * A Map of trackedEntity uid and preheated {@see ProgramInstance}. The
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:178]: This comment is not relevant anymore.
A trackedEntity can be linked to more than one ProgramInstance, there is no validation that will detect it

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EnrollmentInExistingValidationHook.java [hunk 2]
```diff
@@ -104,107 +76,67 @@ public void validateEnrollment( ValidationErrorReporter reporter, Enrollment enr
...
+        Set<Enrollment> dbEnrollment = reporter.getValidationContext().getBundle().getPreheat()
+            .getTrackedEntityToProgramInstanceMap().getOrDefault( enrollment.getTrackedEntity(), new ArrayList<>() )
             .stream()
-            .filter( e -> !e.getEnrollment().equals( enrollment.getEnrollment() )
-                && (EnrollmentStatus.ACTIVE == e.getStatus() || EnrollmentStatus.COMPLETED == e.getStatus()) )
+            .filter( Objects::nonNull )
+            .filter( pi -> pi.getProgram().getUid().equals( program.getUid() )
+                && !pi.getUid().equals( enrollment.getEnrollment() ) )
+            .filter( pi -> ProgramStatus.ACTIVE == pi.getStatus() || ProgramStatus.COMPLETED == pi.getStatus() )
+            .collect( Collectors.toSet() ).stream().map( this::getEnrollmentFromProgramInstance )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EnrollmentInExistingValidationHook.java:101]: Maybe using distinct is less verbose

```suggestion
            .distinct().map( this::getEnrollmentFromProgramInstance )
```

**Review by enricocolasante — APPROVED** [review]
Just 2 small suggestions

**Review by gnespolino — APPROVED** [review]
Good job mate

**Review by ameenhere — APPROVED** [review]

**Pull request merged**